Fix ARMv7 C ABI for Int64s

aoikonomopoulos opened the pull request:
#### 92e73bdff695fa2d3a6250ab4828a845878465b3
<pre>
Fix ARMv7 C ABI for Int64s
<a href="https://bugs.webkit.org/show_bug.cgi?id=279974">https://bugs.webkit.org/show_bug.cgi?id=279974</a>

Reviewed by Justin Michaud.

On ARMv7, Int64 arguments need to be passed on even register pairs (so,
r0/r1 or r2/r3, but not r1/r2). Pass in dummy Int32 arguments in
LowerInt64 so that AirCCallingConvention.cpp does not have to know about
the provenance of Int32 arguments and whether they were originally part
of an Int64.

* Source/JavaScriptCore/b3/B3CCallValue.h:
* Source/JavaScriptCore/b3/B3LowerInt64.cpp:
* Source/JavaScriptCore/b3/air/AirCCallingConvention.cpp:
(JSC::B3::Air::computeCCallingConvention):
(JSC::B3::Air::computeCCallArguments):

Canonical link: <a href="https://commits.webkit.org/284451@main">https://commits.webkit.org/284451@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/7e5d4a3e0d955f20bd346f73dc2ad932be998d84

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/69421 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/48821 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/22094 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/73504 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/20579 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/56622 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/20430 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/55214 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/13678 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/72487 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/44537 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/59937 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/35692 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/41204 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/17367 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/18956 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/62537 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/63145 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/17712 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/75215 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/68667 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/13403 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/16935 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/62881 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/13442 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/60020 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/62787 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/15407 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/10807 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/4421 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/90449 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/44625 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/16045 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/45699 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/46894 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/45440 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

justinmichaud commented on Source/JavaScriptCore/b3/B3LowerInt64.cpp:
```diff
@@ -385,21 +386,36 @@ class LowerInt64 {
                 if (child->type() != Int64)
                     args.append(child);
                 else {
+                    // Offset by one since we add cause we need to pass at least
```
nit

justinmichaud commented on Source/JavaScriptCore/b3/B3LowerInt64.cpp:
```diff
@@ -385,21 +386,36 @@ class LowerInt64 {
+            cCall->setArgContainsLowBitsOfInt64(containsLowBitsOfInt64);
             RELEASE_ASSERT(cCall->numChildren() == 1);
             cCall->effects = m_value->effects();
             cCall->appendArgs(args);
```
Could we just pass in a dummy value instead as padding?

That is, something like Int32, In64, Int64 would become Int32, Const(0), Int64, Int64

It is slightly wasteful, but seems much simpler and wouldn't affect 64-bit at all

aoikonomopoulos replied:
Yah I did consider that, but then LowerInt64.cpp would have more intimate knowledge of the ARMv7 ABI to know where to introduce those dummy values. And AFAICT we'd still need to mark those Const values somehow to avoid emitting needless instructions (though I guess the cost would be in the noise).

justinmichaud commented on Source/JavaScriptCore/b3/B3LowerInt64.cpp:
```diff
@@ -385,21 +386,40 @@ class LowerInt64 {
             if (m_value->type() == Int64) {
-                setMapping(m_value, valueHi(cCall, m_index + 1), valueLo(cCall, m_index + 1));
+                ASSERT(isARM_THUMB2());
+                // Result registers are allocated lo->hi on ARMv7, so undo that here.
```
Can we just make everywhere follow the same ordering? It doesn't really matter which

aoikonomopoulos replied:
So far I've been following the (hi, lo) convention in B3LowerInt64 to remain consistent with the 32-bit-wise 64-bit  arithmetic ops in MacroAssemblerv7.h. Now that we need to massage things for `AirCCallingConvention.cpp`, it may make sense to switch things around. I'm happy to do that in a follow-up patch, since this change would rewrite most of B3LowerInt64, even if mechanically simple.

justinmichaud reviewed (approved):
LGTM. I'm not sure if @Constellation or @kmiller68 have any thoughts.

I would like to see the order of the hi/lo values be the same everywhere, although I won't block this patch on it.

aoikonomopoulos committed:
Fix ARMv7 C ABI for Int64s
https://bugs.webkit.org/show_bug.cgi?id=279974

Reviewed by Justin Michaud.

On ARMv7, Int64 arguments need to be passed on even register pairs (so,
r0/r1 or r2/r3, but not r1/r2). Pass in dummy Int32 arguments in
LowerInt64 so that AirCCallingConvention.cpp does not have to know about
the provenance of Int32 arguments and whether they were originally part
of an Int64.

* Source/JavaScriptCore/b3/B3CCallValue.h:
* Source/JavaScriptCore/b3/B3LowerInt64.cpp:
* Source/JavaScriptCore/b3/air/AirCCallingConvention.cpp:
(JSC::B3::Air::computeCCallingConvention):
(JSC::B3::Air::computeCCallArguments):

Canonical link: https://commits.webkit.org/284451@main
